test/distributed/fsdp/test_wrap.py

```diff
@@ -650,6 +701,116 @@ def test_auto_wrap_with_ignored_modules(self, wrap_method: WrapMethod):
         self.assertTrue(isinstance(model.module[2][0], nn.Linear))
         self.assertTrue(isinstance(model.module[2][1], nn.Linear))
 
+    @unittest.skipIf(torch.cuda.device_count() < 2, "Requires at least 2 GPUs")
+    def test_frozen_params(self):
+        """
+        Tests that mixing frozen/non-frozen parameters in an FSDP instance
+        raises for ``use_orig_params=False`` and warns for ``True``.
+        """
+        for use_orig_params in [True, False]:
+            self._test_frozen_params(use_orig_params)
+
+    def _test_frozen_params(self, use_orig_params: bool):
+        model = LoraModel().cuda()
+        policy = ModuleWrapPolicy({LoraAttention, LoraMLP, LoraDecoder})
+        msg = "layers.0.attn has both parameters with requires_grad=True and False. "
+        if use_orig_params:
+            msg += "We do not recommend wrapping such modules"
+            ctx = self.assertWarnsRegex(UserWarning, msg)
+        else:
+            msg += "FSDP does not support wrapping such modules when use_orig_params=False."
+            ctx = self.assertRaisesRegex(ValueError, msg)
```

Comment:
nit: consider using assertExpectedInline! It makes updating these tests if you tweak the warning/error message much more pleasant, since you can just run the test with `EXPECTTEST_ACCEPT=1` and it will automatically update all the assertions with the new text.